### PR TITLE
spinout enclave-bridge from teerex

### LINKED ISSUE #187: [teerex] separate concerns > enclave_bridge pallet
remove invocation, shielding and confirming stuff from teerex

add new pallet EnclaveBridge

image

align with shard support

### PR DESCRIPTION
**Pull request by brenzi**
closes #187, #199

moved stuff from teerex into a new pallet `pallet_enclave_bridge` and `enclave_bridge_primitives`
* shielding/unshielding
* confirming offchain compute
* everything concerning shards, their config and status
* encountered some brittle testing on the way....fixed it

### CODE REVIEW

File: enclave-bridge/src/tests/test_cases.rs [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+
+		assert_err!(
+			EnclaveBridge::publish_hash(RuntimeOrigin::signed(enclave_signer), hash, vec![], data),
+			Error::<Test>::DataTooLong
+		);
+	})
+}
+
+#[test]
+fn update_ipfs_hash_works() {
```
Comment (clangenb) [inline enclave-bridge/src/tests/test_cases.rs:392]: I don't understand `ipfs_hash` in this context.

File: enclave-bridge/src/mock.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	let teerex_config = pallet_teerex::GenesisConfig { allow_sgx_debug_mode: true };
+	GenesisBuild::<Test>::assimilate_storage(&teerex_config, &mut t).unwrap();
+
+	let mut ext: sp_io::TestExternalities = t.into();
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+/// Run until a particular block.
+#[allow(dead_code)]
```
Comment (clangenb) [inline enclave-bridge/src/mock.rs:164]: why keep this if it is unused?

**Review by clangenb — APPROVED** [review]
Looks good to me in general, only some minors. The splitting up of the PRs does make a lot of sense to me until know!

File: teerex/src/lib.rs [hunk 3]
```diff
@@ -460,7 +326,7 @@ pub mod pallet {
 			Ok(().into())
 		}
 
-		#[pallet::call_index(8)]
+		#[pallet::call_index(4)]
```
[the remark this answers is not in the document]
Comment (brenzi) [reply teerex/src/lib.rs:329]: yes. but I think as this is a breaking change, that's fine

File: enclave-bridge/src/tests/test_cases.rs (hunk 1 above)
Comment (brenzi) [reply enclave-bridge/src/tests/test_cases.rs:392]: me neither. just ported old tests

File: teerex/src/lib.rs [hunk 4]
```diff
@@ -640,47 +450,6 @@ impl<T: Config> Pallet<T> {
...
-		let enclave = Self::sovereign_enclaves(enclave_signer.clone())
-			.ok_or(<Error<T>>::EnclaveIsNotRegistered)?;
-
-		let current_block_number = <frame_system::Pallet<T>>::block_number();
-
-		let new_status = ShardSignerStatus {
-			signer: enclave_signer.clone(),
-			fingerprint: enclave.fingerprint(),
-			last_activity: current_block_number,
-		};
```
[the remark this answers is not in the document]
Comment (brenzi) [reply teerex/src/lib.rs:657]: true. I came to realize we'd have a circular dependency if touch_shard would be used upon registering an enclave. and I think it's not needed....

[the remark this answers is not in the document]
Comment (clangenb) [reply teerex/src/lib.rs:657]: I agree that it is not needed. I think deregistration should only be handled within the teerex pallet. Additional business logic that requires recent activity of the enclaves should be either in the sidechain-pallet of in the layer2 itself.

File: enclave-bridge/src/tests/test_cases.rs (hunk 1 above)
Comment (clangenb) [reply enclave-bridge/src/tests/test_cases.rs:392]: The test is identical to `confirm_processed_parentchain_block_works`, so we should remove this one. and the other one can be renamed to `confirm_processed_parentchain_block_fails_for_not_registered_enclave`.

File: teerex/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brenzi) [reply teerex/src/lib.rs:657]: done

File: enclave-bridge/src/tests/test_cases.rs (hunk 1 above)
Comment (brenzi) [reply enclave-bridge/src/tests/test_cases.rs:392]: I checked. the test names were wrong and one was redundant. fixed

**Review by clangenb — APPROVED** [review]
Thanks, lgtm!

**Pull request merged**